cpp/src/io/json/json_tree.cu

```diff
@@ -254,12 +308,423 @@ tree_meta_t get_tree_representation(device_span<PdaTokenT const> tokens,
                                              is_node);
   CUDF_EXPECTS(parent_node_ids_end - parent_node_ids.begin() == num_nodes,
                "parent node id gather mismatch");
+  nvtxRangePop();
+  nvtxRangePop();  // node_parent_ids
   return {std::move(node_categories),
           std::move(parent_node_ids),
           std::move(node_levels),
           std::move(node_range_begin),
           std::move(node_range_end)};
 }
 
+std::tuple<rmm::device_uvector<NodeIndexT>, rmm::device_uvector<size_type>>
+records_orient_tree_traversal(device_span<SymbolT const> d_input,
+                              tree_meta_t& d_tree,
+                              rmm::cuda_stream_view stream,
+                              rmm::mr::device_memory_resource* mr)
+{
+  CUDF_FUNC_RANGE();
+  /**
+  This algorithm assigns a unique column id to each node in the tree.
+  The row offset is the row index of the node in that column id.
+  Algorithm:
+  1. Convert node_category+fieldname to node_type.
+    a. Create a hashmap to hash field name and assign unique node id as values.
+    b. Convert the node categories to node types.
+       Node type is defined as node category enum value if it is not a field node,
+       otherwise it is the unique node id assigned by the hashmap (value shifted by #NUM_CATEGORY).
+  2. Preprocessing: Translate parent node ids after sorting by level.
+    a. sort by level
+    b. get gather map of sorted indices
+    c. translate parent_node_ids to new sorted indices
+  3. Find level boundaries.
+     copy_if index of first unique values of sorted levels.
+  4. Per-Level Processing: Propagate parent node ids for each level.
+    For each level,
+      a. gather col_id from previous level results. input=col_id, gather_map is parent_indices.
+      b. stable sort by {parent_col_id, node_type}
+      c. scan sum of unique {parent_col_id, node_type}
+      d. scatter the col_id back to stable node_level order (using scatter_indices)
+    Restore original node_id order
+  5. Generate row_offset.
+    a. stable_sort by parent_col_id.
+    b. scan_by_key {parent_col_id} (required only on nodes who's parent is list)
+    c. propagate to non-list leaves from parent list node by recursion
+  **/
+  // GPU version
```

Comment:
```suggestion
```

Reply:
The cpu version is in unit test. (was here originally). will remove this comment.